src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java

```diff
@@ -0,0 +1,188 @@
+    this.namespace = Optional.ofNullable(namespace);
+    this.tableName = Optional.ofNullable(tableName);
+  }
+
+  @Override
+  public void withNamespace(String namespace) {
+    this.namespace = Optional.ofNullable(namespace);
+  }
+
+  @Override
+  public Optional<String> getNamespace() {
+    return namespace;
+  }
+
+  @Override
+  public void withTable(String tableName) {
+    this.tableName = Optional.ofNullable(tableName);
+  }
+
+  @Override
+  public Optional<String> getTable() {
+    return tableName;
+  }
+
+  @Override
+  public Optional<Result> get(Get get) throws ExecutionException {
+    try (Connection connection = dataSource.getConnection()) {
+      return jdbcService.get(get, connection, namespace, tableName);
+    } catch (SQLException e) {
+      throw new ExecutionException("An error occurred", e);
+    }
+  }
+
+  @Override
+  public Scanner scan(Scan scan) throws ExecutionException {
+    Connection connection = null;
+    try {
+      connection = dataSource.getConnection();
+      return jdbcService.scan(scan, connection, namespace, tableName);
+    } catch (SQLException e) {
+      try {
+        if (connection != null) {
+          connection.close();
+        }
+      } catch (SQLException sqlException) {
+        throw new ExecutionException("An error occurred", sqlException);
+      }
+      throw new ExecutionException("An error occurred", e);
+    }
+  }
+
+  @Override
+  public void put(Put put) throws ExecutionException {
+    try (Connection connection = dataSource.getConnection()) {
+      if (!jdbcService.put(put, connection, namespace, tableName)) {
+        throw new NoMutationException("no mutation was applied");
+      }
+    } catch (SQLException e) {
+      throw new ExecutionException("An error occurred", e);
+    }
+  }
+
+  @Override
+  public void put(List<Put> puts) throws ExecutionException {
+    mutate(puts);
+  }
+
+  @Override
+  public void delete(Delete delete) throws ExecutionException {
+    try (Connection connection = dataSource.getConnection()) {
+      if (!jdbcService.delete(delete, connection, namespace, tableName)) {
+        throw new NoMutationException("no mutation was applied");
+      }
+    } catch (SQLException e) {
+      throw new ExecutionException("An error occurred", e);
+    }
+  }
+
+  @Override
+  public void delete(List<Delete> deletes) throws ExecutionException {
+    mutate(deletes);
+  }
+
+  @Override
+  public void mutate(List<? extends Mutation> mutations) throws ExecutionException {
+    Connection connection = null;
+    try {
+      try {
+        connection = dataSource.getConnection();
+        connection.setAutoCommit(false);
+        if (!jdbcService.mutate(mutations, connection, namespace, tableName)) {
+          connection.rollback();
+          throw new NoMutationException("no mutation was applied");
+        } else {
+          connection.commit();
+        }
+      } catch (SQLException e) {
+        if (connection != null) {
```

Comment:
IMO, instead of having nested `try`, it would be better to `try` smaller range and throw with a specific message. What do you think?

```
      try {
        if (connection != null) {
          connection.rollback();
        }
      } catch (SQLException e) {
        throw new ExecutionException("rollback failed.", e);
      }
```